src/Polkadot/SS58Address.h

```diff
@@ -0,0 +1,145 @@
+// Copyright © 2017-2022 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#pragma once
+
+#include "Base58.h"
+#include "Data.h"
+#include "PublicKey.h"
+
+#include <array>
+#include <string>
+
+const std::string SS58Prefix = "SS58PRE";
+
+namespace TW {
+
+/// Base-58-encodeed Substrate address.
+class SS58Address {
+  public:
+    static const size_t checksumSize = 2;
+
+    // networks 0 -- 63 are encoded in one byte
+    static const byte networkSimpleLimit = 0x40;
+    // networks 64 -- 16383 are encoded in 2 bytes, with first byte between 64 and 127
+    static const byte networkFullLimit = 0x80;
+
+    /// Address data consisting of one or more network byte(s) followed by the public key.
+    Data bytes;
+
+    /// Determines whether a string makes a valid address
+    static bool isValid(const std::string& string, uint64_t network) {
+        const auto decoded = Base58::bitcoin.decode(string);
+        byte decodedNetworkSize = 0;
+        uint64_t decodedNetwork = 0;
+        if (!decodeNetwork(decoded, decodedNetworkSize, decodedNetwork)) {
+            return false;
+        }
+        // check size
+        if ((decodedNetworkSize + PublicKey::ed25519Size + checksumSize) != decoded.size()) {
+            return false;
+        }
+        // compare network
+        if (decodedNetwork != network) {
+            return false;
+        }
+        auto checksum = computeChecksum(Data(decoded.begin(), decoded.end() - checksumSize));
+        // compare checksum
+        if (!std::equal(decoded.end() - checksumSize, decoded.end(), checksum.begin())) {
+            return false;
+        }
+        return true;
+    }
+
+    template <typename T>
+    static Data computeChecksum(const T& data) {
+        auto prefix = Data(SS58Prefix.begin(), SS58Prefix.end());
+        append(prefix, Data(data.begin(), data.end()));
+        auto hash = Hash::blake2b(prefix, 64);
+        auto checksum = Data(checksumSize);
+        std::copy(hash.begin(), hash.begin() + checksumSize, checksum.data());
+        return checksum;
+    }
+
+    SS58Address() = default;
+
+    /// Initializes an address with a string representation.
+    SS58Address(const std::string& string, uint64_t network) {
+        if (!isValid(string, network)) {
+            throw std::invalid_argument("Invalid address string");
+        }
+        const auto decoded = Base58::bitcoin.decode(string);
+        bytes.resize(decoded.size() - checksumSize);
+        std::copy(decoded.begin(), decoded.end() - checksumSize, bytes.begin());
+    }
+
+    /// Initializes an address with a public key and network.
+    SS58Address(const PublicKey& publicKey, int32_t network) {
```

Comment:
Let's unify all `network` types, should `int32_t` work other places?
```suggestion
    SS58Address(const PublicKey& publicKey, uint64_t network) {
```

Reply:
Fixed, uint64_t was a mistake